chain/testing.go

```diff
@@ -367,10 +367,12 @@ func (f *Builder) GetTipSet(key types.TipSetKey) (types.TipSet, error) {
 	return types.NewTipSet(blocks...)
 }
 
-// FetchTipSets returns `recur` tipsets from `key` by following parent keys.
-func (f *Builder) FetchTipSets(ctx context.Context, key types.TipSetKey, recur int) ([]types.TipSet, error) {
+// FetchTipSets fetchs the tipset at `tsKey` from the fetchers blockStore backed by the Builder.
+// FetchTipSets will only fetch TipSets whos TipSetKeys evaluate to `false` when passed to `done`,
+// this includes the provided `tsKey`.
+func (f *Builder) FetchTipSets(ctx context.Context, key types.TipSetKey, done func(t types.TipSetKey) bool) ([]types.TipSet, error) {
 	var tips []types.TipSet
-	for i := 0; i < recur; i++ {
+	for !done(key) {
```

Comment:
Testing the key before fetching is slightly awkward, and I think worth changing.  There's no point invoking the predicate on the first `key` passed as parameter to the function: the called wouldn't have called if they didn't want to fetch something. It's more useful to inspect is the tipset that is fetched, with the predicate value indicating whether it wants to fetch another one (that tipset's parent). That will resolve your comment below too.

(This applies to the interface, which should clearly document this semantic).